EpsilonAuxBMS/Src/SetContactorsTask.c

```diff
@@ -0,0 +1,242 @@
+                }
+                else if (!charge || !discharge) // Charge or discharge aren't enabled
+                {
+                    if (!charge) // If charge isn't enabled, turn on charge
+                    {
+                        HAL_GPIO_WritePin(CONTACTOR_ENABLE2_GPIO_Port, CONTACTOR_ENABLE2_Pin, GPIO_PIN_SET);
+                        state = CHARGE_CONTACTOR_ON;
+                    }
+                    else // If discharge isn't enabled, turn on discharge
+                    {
+                        HAL_GPIO_WritePin(CONTACTOR_ENABLE3_GPIO_Port, CONTACTOR_ENABLE3_Pin, GPIO_PIN_SET);
+                        state = DISCHARGE_CONTACTOR_ON;
+                    }
+                }
+
+                break;
+
+            case BLOCKED:
+                if (orionStatus.gpioOk && orionStatus.batteryVoltagesInRange)
+                {
+                    state = prev_state;
+                }
+
+                break;
+
+            default:
+                state = FIRST_CHECK;
+        }
+
+        osMutexRelease(auxStatusMutex);
+    }
+}
+
+int checkContactor(uint16_t pin, GPIO_TypeDef* port, int current_multiplier)
+{
+    float pwr_voltage = 0.0;
+    float pwr_current = 0.0;
+    uint8_t sense_pin;
+    // Get 12bit current analog input
+    uint32_t current_sense = readCurrent();
+
+    if (current_sense == 0xFFFFFFFF)
+    {
+        return 0;
+    }
+
+    pwr_voltage = 3.3 * current_sense / 0xFFF / GAIN; // change ADC value into the voltage read
+    pwr_current = pwr_voltage / CURRENT_SENSE_RESISTOR; // convert voltage to current
+
+    if (pin == 0xFF && port == NULL)
+    {
+        sense_pin = 1;
+    }
+    else
+    {
+        sense_pin = !HAL_GPIO_ReadPin(port, pin);
+    }
+
+    // If current is high for some reason or sense pin is high something is wrong
+    if (pwr_current > CURRENT_LOWER_THRESHOLD * current_multiplier || !sense_pin)
+    {
+        return 0;
+    }
+
+    return 1;
+}
+
+uint32_t readCurrent(void)
+{
+    // Enable current amplifier. Allow time to settle and take multiple valid concurrent readings
+    // to make sure it isn't changing
+    HAL_GPIO_WritePin(CURRENT_SENSE_ENABLE_GPIO_Port, CURRENT_SENSE_ENABLE_Pin, GPIO_PIN_SET);
+    uint32_t sense = 0;
+    uint32_t prev_sense = 0;
+    int counter = 0;
+
+    while (counter < 3)
+    {
+        osDelay(SENSE_SETUP_TIME);
+
+        if (HAL_ADC_PollForConversion(&hadc1, ADC_POLL_TIMEOUT) == HAL_OK)
+        {
+            sense = HAL_ADC_GetValue(&hadc1);
+            int diff = sense - prev_sense;
+
+            if (diff < 0)
+            {
+                diff *= -1;
+            }
+
+            if (diff > 50)
+            {
+                prev_sense = sense;
+                counter = 0;
+            }
+            else
+            {
+                counter++;
+            }
+        }
+        else
+        {
+            sense = 0xFFFFFFFF;
+            break;
+        }
+    }
```

Comment:
Can you explain how this peripheral works as a comment at the top of the func?

Reply:
I should explain how the ADC works or the current amplifier?

Reply:
This function does some pretty weird stuff, why is there multiple attempts and multiple reads?